### PR TITLE
txn: implement CheckSecondaryLocks command

### PR DESCRIPTION
**Pull request by sticnarf**
### What is changed and how it works?

What's Changed:

Add CheckSecondaryLocks to support transaction recovery of async commit.

### Check List 

Tests 

- Unit test

### Release note 

No release note.

### CODE REVIEW

File: src/storage/mvcc/txn.rs [hunk 1]
```diff
@@ -1032,6 +1039,73 @@ impl<S: Snapshot, P: PdClient + 'static> MvccTxn<S, P> {
...
+                let mut need_rollback = true;
+                let status = loop {
+                    if let Some((commit_ts, write)) = self.reader.seek_write(&key, seek_ts)? {
+                        // If the there is a write record whose commit_ts is greater than or equal
+                        // to the start_ts, it is impossible for this key to be locked (restrainted
+                        // by the checks in prewrite and acquire_pessimistic_lock). So we needn't
+                        // write a rollback.
+                        if commit_ts >= start_ts {
+                            need_rollback = false;
+                        }
```
Comment (sticnarf) [inline src/storage/mvcc/txn.rs:1081]: When we support non-unique timestamp, I believe this needs some updating. /cc @MyonKeminta

File: src/server/service/kv.rs [hunk 2]
```diff
@@ -1617,6 +1624,18 @@ txn_command_future!(future_check_txn_status, CheckTxnStatusRequest, CheckTxnStat
             Err(e) => resp.set_error(extract_key_error(&e)),
         }
 });
+txn_command_future!(future_check_secondary_locks, CheckSecondaryLocksRequest, CheckSecondaryLocksResponse, (status, resp) {
+    // All possible errors should be extracted by `extract_region_error`.
```
Comment (youjiali1995) [inline src/server/service/kv.rs:1628]: Does it cover all errors including errors returned by `MvccReader`?

**Comment by youjiali1995** [comment]
> How about `check_secondary_lock` doesn't rollback locks? Because you still have to send requests to roll back all secondary keys even if some keys have been rolled back by `check_secondary_lock`. I think it can be much cleaner.

Oh, one advantage is we can roll back all keys including the primary lock parallelly. If it doesn't rollback locks, we have to roll back the primary lock first, then secondary locks.

**Comment by sticnarf** [comment]
> > How about `check_secondary_lock` doesn't rollback locks? Because you still have to send requests to roll back all secondary keys even if some keys have been rolled back by `check_secondary_lock`. I think it can be much cleaner.
> 
> Oh, one advantage is we can roll back all keys including the primary lock parallelly. If it doesn't rollback locks, we have to roll back the primary lock first, then secondary locks.

I think we must roll back pessimistic locks.

If we don't, the client will just get some pessimistic locks. Then, what should the client do? It cannot roll back the primary lock because at this time another client might succeed to prewrite the secondary locks and report success to the user.
If the client needs to roll back the secondary locks, it doesn't have a good command to do it as well. It cannot use `rollback` because it may roll back a prewritten lock by mistake. A `pessimistic_rollback` may be also overriden by a re-sent stale `acquire_pessimistic_lock`, making us hard to determine the status of the lock.

That's the reason why I choose to rollback here...

File: src/server/service/kv.rs (hunk 2 above)
Comment (sticnarf) [reply src/server/service/kv.rs:1628]: Ah, I thought the errors come from the engine and they were covered. Now I find bad write record format also has an error.  So a key error is necessary... https://github.com/pingcap/kvproto/pull/657/files

File: src/storage/mvcc/txn.rs (hunk 1 above)
Comment (sticnarf) [reply src/storage/mvcc/txn.rs:1081]: Could you clarify what's the relationship with #7989?
I think this optimization is something not necessary but nice to have.

Comment (sticnarf) [reply src/storage/mvcc/txn.rs]: Oh, I find we this optimization is not sound. Pessimistic locks can get around it.

**Comment by sticnarf** [comment]
@youjiali1995 @MyonKeminta @cfzjywxk Could you take a look again? Thanks!

File: src/storage/mvcc/txn.rs [hunk 3]
```diff
@@ -1032,6 +1039,48 @@ impl<S: Snapshot, P: PdClient + 'static> MvccTxn<S, P> {
...
+                Some((commit_ts, write_type)) if write_type != WriteType::Rollback => {
+                    SecondaryLockStatus::Committed(commit_ts)
+                }
+                _ => SecondaryLockStatus::RolledBack,
+            },
+        };
+        // We must protect this rollback in case this rollback is collapsed and a stale
+        // acquire_pessimistic_lock and prewrite succeed again.
+        let write = Write::new_rollback(self.start_ts, true);
+        self.put_write(key.clone(), self.start_ts, write.as_ref().to_bytes());
```
Comment (cfzjywxk) [inline src/storage/mvcc/txn.rs:1077]: This rollback write should be done checking the `status` variable?

Comment (sticnarf) [reply src/storage/mvcc/txn.rs]: Ah, sorry. I delete `need_rollback` by mistake when reorganizing the code...
We shoudn't write rollback when the write record exists.

Comment (sticnarf) [reply src/storage/mvcc/txn.rs]: I update this PR and add one more test for it.

File: src/storage/mvcc/txn.rs [hunk 4]
```diff
@@ -1032,6 +1039,61 @@ impl<S: Snapshot, P: PdClient + 'static> MvccTxn<S, P> {
...
+        start_ts: TimeStamp,
+    ) -> Result<(SecondaryLockStatus, Option<ReleasedLock>)> {
+        let mut released_lock = None;
+        let (status, need_rollback) = match self.reader.load_lock(&key)? {
+            Some(lock) if lock.ts == start_ts => {
+                let status = if lock.lock_type == LockType::Pessimistic {
+                    released_lock = self.unlock_key(key.clone(), true);
+                    SecondaryLockStatus::RolledBack
+                } else {
+                    SecondaryLockStatus::Locked(lock)
```
Comment (cfzjywxk) [inline src/storage/mvcc/txn.rs:1064]: In this branch, if the corresponding prewrite lock exists, `need_rollback` flag should be false? Maybe we could add test cases to check flags or rollback records.

File: components/tikv_alloc/src/lib.rs [hunk 5]
```diff
@@ -82,6 +82,7 @@
 #[macro_use]
 extern crate log;
 
+#[cfg(feature = "jemalloc")]
 #[macro_use]
 extern crate lazy_static;
```
Comment (MyonKeminta) [inline components/tikv_alloc/src/lib.rs:87]: why need this?

File: src/storage/mvcc/txn.rs (hunk 4 above)
Comment (sticnarf) [reply src/storage/mvcc/txn.rs]: ...sorry 😷
I wrote a wrong assertion in the test... I've fixed it and add assertions after each `check_secondary` operation.

File: components/tikv_alloc/src/lib.rs (hunk 5 above)
Comment (sticnarf) [reply components/tikv_alloc/src/lib.rs:87]: A latest update in master bring a `unused macro_use` warning.

File: src/server/service/kv.rs [hunk 6]
```diff
@@ -1617,6 +1624,18 @@ txn_command_future!(future_check_txn_status, CheckTxnStatusRequest, CheckTxnStat
...
         }
 });
+txn_command_future!(future_check_secondary_locks, CheckSecondaryLocksRequest, CheckSecondaryLocksResponse, (status, resp) {
+    match status {
+        Ok(SecondaryLocksStatus::Locked(locks)) => {
+            resp.set_locks(locks.into());
+        },
+        Ok(SecondaryLocksStatus::Committed(ts)) => {
+            resp.set_commit_ts(ts.into_inner());
+        },
```
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply src/server/service/kv.rs:1634]: Oh never mind

File: src/storage/mvcc/txn.rs [hunk 7]
```diff
@@ -1032,6 +1039,48 @@ impl<S: Snapshot, P: PdClient + 'static> MvccTxn<S, P> {
...
+    pub fn check_secondary_lock(
+        &mut self,
+        key: &Key,
+        start_ts: TimeStamp,
+    ) -> Result<(SecondaryLockStatus, Option<ReleasedLock>)> {
+        let mut released_lock = None;
+        let status = match self.reader.load_lock(&key)? {
+            Some(lock) if lock.ts == start_ts => {
+                if lock.lock_type == LockType::Pessimistic {
+                    released_lock = self.unlock_key(key.clone(), true);
```
[the remark this answers is not in the document]
Comment (sticnarf) [reply src/storage/mvcc/txn.rs:1061]: This line is to remove the LOCF_CF record. A protected rollback is written here: https://github.com/tikv/tikv/pull/8337/files/95afa049ef6b93552a582be8f379da008bd4d79c#diff-10ac7ebd80ca4da7cccc8e3539f12eb4R1076

File: src/server/service/kv.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (sticnarf) [reply src/server/service/kv.rs:1634]: If any of the secondary locks is committed, the primary lock must have been committed. So I think it's ok.

**Comment by cfzjywxk** [comment]
LGTM

**Pull request merged**